### PR TITLE
Lazily figuring out designated node

### PR DESCRIPTION
**Pull request by wavesoft**
## High-level description

This fixes a potential issue when upgrading  `etcd` after the https://github.com/dcos/dcos/pull/7313#issue-423136936 fix was landed.

This commit covers a corner case where all the `etcd` instances are restarted, or if the cluster quorum is not yet established before the `dcos-etcd` service is started. 

Namely, the logic that detects the `etcd` node to use for the CLI commands is now moved away from the `EtcdctlHelper` constructor and it's lazily evaluated when needed.

## Corresponding DC/OS tickets (required)

  - [D2IQ-69069](https://jira.d2iq.com/browse/D2IQ-69069) etcd failed to upgrade on clusters with COPS-6183 fix

## Related tickets (optional)

  - [D2IQ-ID](https://jira.mesosphere.com/browse/D2IQ-<number>) JIRA title / short description.

### CODE REVIEW

**Comment by wavesoft** [comment]
@mesosphere-mergebot request-review @vespian @jongiddy

File: packages/etcd/extra/etcd_discovery/etcd_discovery.py [hunk 1]
```diff
@@ -152,7 +152,7 @@ def zk_lock(zk: KazooClient, lock_path: str, contender_id: str,
     lock = zk.Lock(lock_path, contender_id)
     try:
         log.info("Acquiring ZooKeeper lock.")
-        lock.acquire(blocking=True, timeout=timeout)
+        lock.acquire(blocking=True, timeout=timeout, ephemeral=True)
```
Comment (wavesoft) [inline packages/etcd/extra/etcd_discovery/etcd_discovery.py:155]: Redundant default, added for clarity.

File: gen/dcos-config.yaml [hunk 2]
```diff
@@ -1926,6 +1926,14 @@ package:
               {
                   "Command": ["/opt/mesosphere/bin/dcos-shell", "dcos-etcdctl", "diagnostic"],
                   "Role": ["master"]
+              },
+              {
+                  "Command": ["/bin/sh", "-c", "source /run/dcos/etc/etcd/zk-creds; echo -e \"addauth digest ${DATASTORE_ZK_USER}:${DATASTORE_ZK_SECRET}\\nget /etcd/nodes\" | /opt/mesosphere/bin/dcos-shell zkCli.sh 2>&1"],
```
Comment (vinodkone) [inline gen/dcos-config.yaml:1931]: are these variables accessible from this file?

File: gen/dcos-config.yaml [hunk 3]
```diff
@@ -1926,6 +1926,14 @@ package:
...
+                  "Role": ["master"]
+              },
+              {
+                  "Command": ["/bin/sh", "-c", "source /run/dcos/etc/etcd/zk-creds; echo -e \"addauth digest ${DATASTORE_ZK_USER}:${DATASTORE_ZK_SECRET}\\nls /etcd/locking\" | /opt/mesosphere/bin/dcos-shell zkCli.sh 2>&1"],
```
Comment (vinodkone) [inline gen/dcos-config.yaml:1935]: can you also include the state files in the bundle, as discussed?

File: packages/etcd/README.md [hunk 4]
```diff
@@ -55,3 +55,57 @@ The above command typically presents the results of etcdctl command [`endpoint h
 
 - `endpoint health` checks the healthiness of on the current etcd instance, which should meet `healthy` 
 - `member list -w json` returns the cluster members, which should return all etcd instances
+
+## Recovering an etcd cluster
+
+If you ever encounter a situation where you loose the majority of your etcd quorum, you might still be able to recover your data from just a single healthy node using the following procedure.
```
Comment (vinodkone) [inline packages/etcd/README.md:61]: lose*

File: gen/dcos-config.yaml (hunk 2 above)
Comment (wavesoft) [reply gen/dcos-config.yaml:1931]: Yes, I also verified it works by invoking the same cmdline from the masters.

File: gen/dcos-config.yaml (hunk 3 above)
Comment (wavesoft) [reply gen/dcos-config.yaml:1935]: Whoops, true. Will add

**Comment by vinodkone** [comment]
@mesosphere-mergebot backport  2.1

**Pull request merged**